fixedobject.go

```diff
@@ -0,0 +1,267 @@
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"io"
+	"reflect"
+	"strconv"
+)
+
+type ObjectField struct {
+	Aliases []string
+	Schema  Schema
+}
+
+type FixedObjectSchema struct {
+	SchemaOptions
+	Fields []ObjectField
+}
+
+func (s *FixedObjectSchema) DefaultGOType() reflect.Type {
+	type tmpStruct struct {
+	}
+	var t tmpStruct
+	return reflect.TypeOf(t)
+}
+
+func (s *FixedObjectSchema) MarshalJSON() ([]byte, error) {
+	tmpMap := make(map[string]interface{}, 3)
+	tmpMap["type"] = "object"
+	tmpMap["nullable"] = strconv.FormatBool(s.SchemaOptions.Nullable)
+
+	var fieldMap []map[string]interface{}
+
+	for i := range s.Fields {
+		fields := make(map[string]interface{})
+		fields["name"] = s.Fields[i].Aliases[0]
+
+		b, err := s.Fields[i].Schema.MarshalJSON()
+		if err != nil {
+			return nil, err
+		}
+		err = json.Unmarshal(b, &fields)
+		if err != nil {
+			return nil, err
+		}
+		fieldMap = append(fieldMap, fields)
+	}
+
+	tmpMap["fields"] = fieldMap
+	return json.MarshalIndent(tmpMap, "", "   ")
+}
+
+// Bytes encodes the schema in a portable binary format
+func (s *FixedObjectSchema) MarshalSchemer() []byte {
+
+	// fixedObject schemas are 1 byte long
+	var schemaBytes []byte = []byte{0b00101001}
+
+	// The most signifiant bit indicates whether or not the type is nullable
+	if s.SchemaOptions.Nullable {
+		schemaBytes[0] |= 128
+	}
+
+	// encode total number of fields as a varint
+	buf := make([]byte, binary.MaxVarintLen64)
+	varIntByteLength := binary.PutVarint(buf, int64(len(s.Fields)))
+
+	schemaBytes = append(schemaBytes, buf[0:varIntByteLength]...)
+
+	// also need to concatenate the schemas for all other fields
+	for _, f := range s.Fields {
+
+		// first encode the number of aliases for this field
+		// (which will always be at least one... meaning the name of the field from the source struct at least!)
+		buf := make([]byte, binary.MaxVarintLen64)
+		varIntByteLength := binary.PutVarint(buf, int64(len(f.Aliases)))
+
+		schemaBytes = append(schemaBytes, buf[0:varIntByteLength]...)
+
+		// now write each field alias
+		for i := 0; i < len(f.Aliases); i++ {
+			s := f.Aliases[i]
+			varLenStringSchema := SchemaOf(s)
+			var buf bytes.Buffer
+			varLenStringSchema.Encode(&buf, s)
+			schemaBytes = append(schemaBytes, buf.Bytes()...)
+		}
+
+		schemaBytes = append(schemaBytes, f.Schema.MarshalSchemer()...)
+	}
+
+	return schemaBytes
+}
+
+// Encode uses the schema to write the encoded value of v to the output stream
+func (s *FixedObjectSchema) Encode(w io.Writer, i interface{}) error {
+
+	v := reflect.ValueOf(i)
+
+	ok, err := PreEncode(s, w, &v)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		return nil
+	}
+
+	t := v.Type()
+	k := t.Kind()
+
+	if k != reflect.Struct {
+		return fmt.Errorf("fixedObjectSchema can only encode structs")
+	}
+	// loop through all the schemas in this object
+	// and encode each field
+	for i := 0; i < len(s.Fields); i++ {
+
+		f := v.Field(i)
+		err := s.Fields[i].Schema.Encode(w, f.Interface())
```

Comment:
Use `EncodeValue`

Reply:
One thing to note here is that we don't have a function EncodeValue(). (We have DecodeValue() only)